posts/2023-07-18-lazy-layout.lhs

```diff
@@ -0,0 +1,589 @@
+a sort of restricted affine 2D [transformation matrix].
+However, it's not immediately clear that this is a valid Semigroup, so we will
+be rigorous and provide a proof that `a <> (b <> c) == (a <> b) <> c`.
+
+[transformation matrix]: https://en.wikipedia.org/wiki/Transformation_matrix
+
+<details><summary>Proof of associativity...</summary>
+
+~~~~~{.haskell}
+Tr ax ay as <> (Tr bx by bs <> Tr cx cy cs)
+
+-- Definition of <>
+= Tr (ax * (bs * cs) + (bx * cs + cx))
+     (ay * (bs * cs) + (by * cs + cy))
+     (as * (bs * cs))
+
+-- Associativity of * and +
+= Tr (ax * bs * cs + bx * cs + cx)
+     (ay * bs * cs + by * cs + cy)
+     ((as * bs) * cs)
+
+-- Distributivity of * over +
+= Tr ((ax * bs + bx) * cs + cx)
+     ((ay * bs + by) * cs + cy)
+     ((as * bs) * cs)
+
+-- Definition of <>
+= (Tr ax ay as <> T b by bs) <> Tr cx cy cs
+~~~~~~
+
+</details>
+
+Aside from the composition operator `<>`, we also need to provide an identity
+`Transform`, which is just offsetting by 0 and scaling by 1:
+
+> instance Monoid Transform where
+>   mempty = Tr 0 0 1
+
+Proving that the identity holds on `mempty` is simple so we will only do one
+side, namely `a <> mempty == a`.
+
+<details><summary>Proof of Monoid right identity...</summary>
+
+~~~~~{.haskell}
+Tr ax ay as <> mempty
+
+-- Definition of mempy
+= Tr ax ay as <> Tr 0 0 1
+
+-- Definition of <>
+= Tr (ax * 1 + 0) (ay * 1 + 0) (as * 1)
+
+-- Cancellative property of 0 over +
+-- Identity of 1 over *
+= Tr ax ay as
+~~~~~
+
+</details>
+
+The lazy layout
+===============
+
+Our main `layoutCollage` function takes the user-specified tree as input,
+and annotates each element with a `Transform`.
+In addition to that, we also produce the `Size` of the final image so we can
+allocate space for it.
+
+> layoutCollage
+>   :: Sized img
+>   => Collage img
+>   -> (Collage (img, Transform), Size)
+
+All `layoutCollage` does is call `layout` --- our _circular_ program --- with
+the identity transformation:
+
+> layoutCollage = layout mempty
+
+`layout` takes the size and position of the current element as an argument,
+and determines the sizes and positions of a tree recursively.
+
+There are some similarities with the algorithms present in browser engines,
+where a parent element will first lay out its children, and then use their
+properties to determine its own width.
+
+However, we will use Haskell's laziness to do this in a single top-down pass.
+We provide a declarative algorithm and we leave the decision about what to
+calculate when --- more concretely, propagating the requested sizes of the
+children back up the tree before constructing the transformations --- to the
+compiler!
+
+> layout
+>   :: Sized img
+>   => Transform
+>   -> Collage img
+>   -> (Collage (img, Transform), Size)
+
+Placing a single image is easy, since we are passing in the transformation,
+and we return its _requested_ size which is just the original size of the image.
```

Comment:
The and fells off here, probably because it doesn't relate to `placing a single image`.

```suggestion
Placing a single image is easy, we simply pass in the transformation.
We return the _requested_ size - which is just the original size of the image.
```